File: multi-gammon-core/src/main/java/org/akoshterek/backgammon/agent/PubEvalAgent.java
package org.akoshterek.backgammon.agent;

import org.akoshterek.backgammon.Constants;
import org.akoshterek.backgammon.board.Board;
import org.akoshterek.backgammon.board.PositionClass;
import org.akoshterek.backgammon.eval.Reward;

import java.nio.file.Path;

/**
 * @author Alex
 *         date 31.08.2015.
 */
public class PubEvalAgent extends AbsAgent {
    private final PubEval eval = new PubEval();

    public PubEvalAgent(Path path) {
        super(path);
        fullName = "PubEval";
    }

    @Override
    public Reward evalContact(Board board) {
        Reward reward = new Reward();
        int [] pos = new int[28];
        preparePos(board, pos);

        int race = curPC.getValue() <= PositionClass.CLASS_RACE.getValue() ? 1 : 0;
        reward.data[Constants.OUTPUT_WIN] = eval.pubeval(race, pos);
        return reward;
    }

    private void preparePos(Board board, int[] pos)  {
        Board tmpBoard = new Board(board);
        tmpBoard.swapSides();

        int[] men = new int[2];
        tmpBoard.chequersCount(men);

        for(int i = 0; i < 24; i++) {
            pos[i+1] = tmpBoard.anBoard[Board.SELF][i];
            if(tmpBoard.anBoard[Board.OPPONENT][23 - i] != 0) {
                pos[i + 1] = -tmpBoard.anBoard[Board.OPPONENT][23 - i];
            }
        }

        pos[25] = tmpBoard.anBoard[Board.SELF][Board.BAR];
        pos[0] = -tmpBoard.anBoard[Board.OPPONENT][Board.BAR];
        pos[26] = 15 - men[Board.SELF];
        pos[27] = -(15 - men[Board.OPPONENT]);
    }
}

File: multi-gammon-core/src/main/java/org/akoshterek/backgammon/agent/pubeval/PubEvalAgent.scala
package org.akoshterek.backgammon.agent.pubeval

import java.nio.file.Path

import org.akoshterek.backgammon.Constants
import org.akoshterek.backgammon.agent.AbsAgent
import org.akoshterek.backgammon.board.{Board, PositionClass}
import org.akoshterek.backgammon.eval.Reward

object PubEvalAgent {
  def apply(path: Path): PubEvalAgent = {
    new PubEvalAgent(path, PubEvalDefaultWeights.contactWeights, PubEvalDefaultWeights.raceWeights)
  }
}

class PubEvalAgent(override val path: Path, val contactWeights: Array[Float], val raceWeights: Array[Float])
  extends AbsAgent("PubEval", path) {

  private val eval: PubEval = new PubEval(contactWeights, raceWeights)

  def evalContact(board: Board): Reward = evaluate(board)

  override def copyAgent(): PubEvalAgent = {
    new PubEvalAgent(path, contactWeights, raceWeights)
  }

  private def evaluate(board: Board): Reward = {
    val reward = Reward.rewardArray[Float]
    val pos: Array[Int] = new Array[Int](28)
    preparePos(board, pos)

    val race: Int = if (curPC.getValue <= PositionClass.CLASS_RACE.getValue) 1 else 0
    reward(Constants.OUTPUT_WIN) = eval.evaluate(race, pos).toFloat
    new Reward(reward)
  }

  private def preparePos(board: Board, pos: Array[Int]): Unit = {
    val (opponent, self) = board.chequersCount

    for (i <- 0 until Board.HALF_BOARD_SIZE - 1) {
      pos(i + 1) = board.anBoard(Board.SELF)(i)
      if (board.anBoard(Board.OPPONENT)(23 - i) != 0) {
        pos(i + 1) = -board.anBoard(Board.OPPONENT)(23 - i)
      }
    }

    pos(25) = board.anBoard(Board.SELF)(Board.BAR)
    pos(0) = -board.anBoard(Board.OPPONENT)(Board.BAR)
    pos(26) = 15 - self
    pos(27) = -(15 - opponent)
  }

  override def clone: PubEvalAgent = {
    new PubEvalAgent(path, eval.contactWeights, eval.raceWeights)
  }
}